core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -301,6 +303,116 @@ class TransactionStateManager(brokerId: Int,
+      // the append response should only contain the topics partition
+      if (responseStatus.size != 1 || !responseStatus.contains(topicPartition))
+        throw new IllegalStateException("Append status %s should only have one partition %s"
+          .format(responseStatus, topicPartition))
+
+      val status = responseStatus(topicPartition)
+
+      var responseError = if (status.error == Errors.NONE) {
+        Errors.NONE
+      } else {
+        debug(s"Transaction state update $txnMetadata for $transactionalId failed when appending to log " +
+          s"due to ${status.error.exceptionName}")
+
+        // transform the log append error code to the corresponding coordinator error code
+        status.error match {
+          case Errors.UNKNOWN_TOPIC_OR_PARTITION
+               | Errors.NOT_ENOUGH_REPLICAS
+               | Errors.NOT_ENOUGH_REPLICAS_AFTER_APPEND
+               | Errors.REQUEST_TIMED_OUT => // note that for timed out request we return NOT_AVAILABLE error code to let client retry
+
+            debug(s"Appending transaction message $txnMetadata for $transactionalId failed due to " +
+              s"${status.error.exceptionName}, returning ${Errors.COORDINATOR_NOT_AVAILABLE} to the client")
+
+            Errors.COORDINATOR_NOT_AVAILABLE
+
+          case Errors.NOT_LEADER_FOR_PARTITION =>
+
+            debug(s"Appending transaction message $txnMetadata for $transactionalId failed due to " +
+              s"${status.error.exceptionName}, returning ${Errors.NOT_COORDINATOR} to the client")
+
+            Errors.NOT_COORDINATOR
+
+          case Errors.MESSAGE_TOO_LARGE
+               | Errors.RECORD_LIST_TOO_LARGE
+               | Errors.INVALID_FETCH_SIZE =>
+
+            error(s"Appending transaction message $txnMetadata for $transactionalId failed due to " +
+              s"${status.error.exceptionName}, returning UNKNOWN error code to the client")
+
+            Errors.UNKNOWN
+
+          case other =>
+            error(s"Appending metadata message f$txnMetadata or $transactionalId failed due to " +
+              s"unexpected error: ${status.error.exceptionName}")
+
+            other
+        }
+      }
+
+      // now try to update the cache: we need to update the status in-place instead of
+      // overwritting the whole object to ensure synchronization
+      if (responseError == Errors.NONE) {
+        getTransaction(transactionalId) match {
+          case Some(metadata) =>
+            metadata synchronized {
+              if (metadata.pid != txnMetadata.pid ||
+                metadata.epoch != txnMetadata.epoch ||
+                metadata.txnTimeoutMs != txnMetadata.txnTimeoutMs ||
+                !metadata.transitionTo(txnMetadata.state)) {
+                // the record has already been appended but the cache has been modified.
+                // return PRODUCER_FENCED to let the client to close itself
+                responseError = Errors.PRODUCER_FENCED
```

Comment:
I don't think all of these conditions should return `PRODUCER_FENCED`, only if the pid or epoch doesn't match. 

Most importantly, if the state has changed to the desired state already, this could be a spurious retry and should not be an error.